File: python/test/t_TNC_nonlinear.py
#! /usr/bin/env python

from __future__ import print_function
from openturns import *
from math import *

TESTPREAMBLE()


def printNumericalPoint(point, digits):
    oss = "["
    eps = pow(0.1, digits)
    format = "%." + str(digits) + "f"
    for i in range(point.getDimension()):
        if i == 0:
            sep = ""
        else:
            sep = ","
        if fabs(point[i]) < eps:
            oss += sep + format % fabs(point[i])
        else:
            oss += sep + format % point[i]
        sep = ","
    oss += "]"
    return oss

try:
    levelFunction = NumericalMathFunction(
        ["x1", "x2", "x3", "x4"], ["y1"], ["x1*cos(x1)+2*x2*x3-3*x3+4*x3*x4"])
    specific = TNCSpecificParameters()
    startingPoint = NumericalPoint(4, -0.01)

    startingPointNearMinimizationCorner = NumericalPoint(4)
    startingPointNearMinimizationCorner[0] = 3.0
    startingPointNearMinimizationCorner[0] = -2.5
    startingPointNearMinimizationCorner[0] = 4.5
    startingPointNearMinimizationCorner[0] = -2.5
    startingPointNearMaximizationCorner = NumericalPoint(4)
    startingPointNearMaximizationCorner[0] = -2.5
    startingPointNearMaximizationCorner[1] = 4.5
    startingPointNearMaximizationCorner[2] = 4.5
    startingPointNearMaximizationCorner[3] = 4.5

    bounds = Interval(NumericalPoint(4, -3.0), NumericalPoint(4, 5.0))
    myAlgorithm = TNC(
        specific, levelFunction, bounds, startingPointNearMinimizationCorner,
        BoundConstrainedAlgorithmImplementationResult.MINIMIZATION)
    myAlgorithm.run()
    print("minimizer = ", printNumericalPoint(myAlgorithm.getResult().getOptimizer(), 4),
          " value=%.5e" % myAlgorithm.getResult().getOptimalValue())

    myAlgorithm = TNC(
        specific, levelFunction, bounds, startingPointNearMaximizationCorner,
        BoundConstrainedAlgorithmImplementationResult.MAXIMIZATION)
    myAlgorithm.run()
    print("maximizer = ", printNumericalPoint(myAlgorithm.getResult().getOptimizer(), 4),
          " value=%.5e" % myAlgorithm.getResult().getOptimalValue())

except:
    import sys
    print("t_TNC_nonlinear.py", sys.exc_info()[0], sys.exc_info()[1])

File: lib/test/t_TNC_nonlinear.cxx
//                                               -*- C++ -*-
/**
 *  @brief The test file of class TNC for solving a linear problem
 *
 *  Copyright 2005-2015 Airbus-EDF-IMACS-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "OT.hxx"
#include "OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

String printNumericalPoint(const NumericalPoint & point, const UnsignedInteger digits)
{
  OSS oss;
  oss << "[";
  NumericalScalar eps(pow(0.1, 1.0 * digits));
  for (UnsignedInteger i = 0; i < point.getDimension(); i++)
  {
    oss << std::fixed << std::setprecision(digits) << (i == 0 ? "" : ",") << Bulk<double>((fabs(point[i]) < eps) ? fabs(point[i]) : point[i]);
  }
  oss << "]";
  return oss;
}

int main(int argc, char *argv[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);

  try
  {

      // Test function operator ()
      Description input(4);
      input[0] = "x1";
      input[1] = "x2";
      input[2] = "x3";
      input[3] = "x4";
      NumericalMathFunction levelFunction(input, Description(1, "y1"), Description(1, "x1*cos(x1)+2*x2*x3-3*x3+4*x3*x4"));
      TNCSpecificParameters specific;
      NumericalPoint startingPointNearMinimizationCorner(4);
      startingPointNearMinimizationCorner[0] = 3.0;
      startingPointNearMinimizationCorner[0] = -2.5;
      startingPointNearMinimizationCorner[0] = 4.5;
      startingPointNearMinimizationCorner[0] = -2.5;
      NumericalPoint startingPointNearMaximizationCorner(4);
      startingPointNearMaximizationCorner[0] = -2.5;
      startingPointNearMaximizationCorner[1] = 4.5;
      startingPointNearMaximizationCorner[2] = 4.5;
      startingPointNearMaximizationCorner[3] = 4.5;

      Interval bounds(NumericalPoint(4, -3.0), NumericalPoint(4, 5.0));
      {
        TNC myAlgorithm(specific, levelFunction, bounds, startingPointNearMinimizationCorner, TNC::Result::MINIMIZATION);
        myAlgorithm.run();
        fullprint << "minimizer = " << printNumericalPoint(myAlgorithm.getResult().getOptimizer(), 4) << " value=" << myAlgorithm.getResult().getOptimalValue() << std::endl;
      }
      {
        TNC myAlgorithm(specific, levelFunction, bounds, startingPointNearMaximizationCorner, TNC::Result::MAXIMIZATION);
        myAlgorithm.run();
        fullprint << "maximizer = " << printNumericalPoint(myAlgorithm.getResult().getOptimizer(), 4) << " value=" << myAlgorithm.getResult().getOptimalValue() << std::endl;
      }
  }
  catch (TestFailed & ex)
  {
    std::cerr << ex << std::endl;
    return ExitCode::Error;
  }

  return ExitCode::Success;
}
